auth-gateway/index.js

```diff
@@ -0,0 +1,162 @@
+require('dotenv').config();
+
+const polka = require('polka');
+const jwksRsa = require('jwks-rsa');
+const jwt = require('jsonwebtoken');
+
+const { AUTH0_DOMAIN, AUTH0_AUDIENCE } = process.env;
+const AUTH0_ISSUER = `https://${AUTH0_DOMAIN}/`;
+
+// Grabs public key, caches it, from Auth0's servers
+// asymmetric signature used to ensure that not only the jwt was unmodified
+// since signed, but that the issuer was Auth0 (since they hold the private key)
+const jwksFn = jwksRsa({
+  cache: true,
+  rateLimit: true,
+  jwksRequestsPerMinute: 5,
+  jwksUri: `https://${AUTH0_DOMAIN}/.well-known/jwks.json`
+});
+
+// Ref: https://github.com/auth0/node-jwks-rsa/blob/master/src/integrations/express.js
+function getSecretKey(header, cb) {
+  jwksFn.getSigningKey(header.kid, (err, key) => {
+    if (err) {
+      cb(err, null);
+      return;
+    }
+
+    cb(null, key.publicKey || key.rsaPublicKey);
+  });
+}
+
+// Verifies signature, as well as the issuer and audience, to guarantee
+// that the token was truly for our resource server
+const verifyToken = token =>
+  new Promise((resolve, reject) => {
+    jwt.verify(
+      token,
+      getSecretKey,
+      {
+        issuer: AUTH0_ISSUER,
+        audience: AUTH0_AUDIENCE,
+        clockTolerance: 2, // seconds to deal with clock skew
+        algorithms: ['RS256']
+      },
+      (err, dToken) => {
+        if (err) {
+          console.error(err);
+          return reject(err);
+        }
+
+        return resolve(dToken);
+      }
+    );
+  });
+
+const bearerPrefix = 'Bearer ';
+const bearerPrefixLen = bearerPrefix.length;
+
+const cookiePrefix = 'access_token=';
+const cookieOffset = cookiePrefix.length;
+
+const getAuthToken = req => {
```

Comment:
Seems vaguely not great to put a bearer token in the URL, even if its encrypted during transport. It seems not impossible that it might get inadvertently written into a log somewhere.

Reply:
We use 2, cookie and header.

I agree that the token shouldn't be sent in the URL

Supporting cookies is the least-contrived way I can think of to authorize every Jupyter request. After the initial redirect Jupyter fires off a huge number of these. We don't control those requests, and while we could contrive some way to add state to these requests, there is already a standard way of doing this, a cookie.

Reply:
Also: https://tools.ietf.org/html/rfc6750


2.3.  URI Query Parameter

   When sending the access token in the HTTP request URI, the client
   adds the access token to the request URI query component as defined
   by "Uniform Resource Identifier (URI): Generic Syntax" [RFC3986],
   using the "access_token" parameter.

   For example, the client makes the following HTTP request using
   transport-layer security:

     GET /resource?access_token=mF_9.B5f-4.1JqM HTTP/1.1
     Host: server.example.com

   The HTTP request URI query can include other request-specific
   parameters, in which case the "access_token" parameter MUST be
   properly separated from the request-specific parameters using "&"
   character(s) (ASCII code 38).

Reply:
I added back URI token support, because I don't think this endpoint should make decisions about spec-compliant uses.

Reply:
The same section notes that servers may choose to not support URI auth.

>    Because of the security weaknesses associated with the URI method
   (see Section 5), including the high likelihood that the URL
   containing the access token will be logged, it SHOULD NOT be used
   unless it is impossible to transport the access token in the
   "Authorization" request header field or the HTTP request entity-body.
   Resource servers MAY support this method.

I'm inclined to not support it for the reasons noted at the end of section 2.3.

Reply:
Sure